Contextual menu for multi selecting episodes

peakvalleytech opened the pull request:
<!-- Please make sure that you have read our contribution guidelines: https://github.com/AntennaPod/AntennaPod/blob/develop/CONTRIBUTING.md#submit-a-pull-request -->
Here is a working build I did implementing your mockup: https://drive.google.com/file/d/1A0fMp7SemlpHHltod-x0MQsDRn9TZDwg/view

I'm not close to done, but I wanted to get your feedback on it before I proceed further.

What I added:
* Integrated multi select on same existing episodes screen using contextual menu (https://developer.android.com/guide/topics/ui/menus.html#context-menu)
* Moved speed dial to episodes screen
* Ported existing multi select actions to existing screen
* Have NOT added new multi select actions

ByteHamster commented:
Oh, I thought you would do the subscriptions screen, not the episodes screen :) I have not tried it yet but I skimmed the code. Some thoughts:

- I think it would be good to still show the context menu on long press. Only when clicking a new "select" entry there (or, alternatively, when clicking it in the 3-dot-menu), then enter the ActionMode. Reason is that it would otherwise make the more common action (changing a single episode) need a lot more clicks.
- When having the ActionMode, I don't think we need the FAB anymore. The actions can directly be added to the ActionMode.

peakvalleytech commented:
idk, it seems odd to have it has an option in the context menu. How about adding a 3-dot like button next to the play button. That is very common UI pattern. With this approach, we will be implementing 2 common patterns versus 1. It seems like a very good choice to me. 

Regarding the FAB. I think with our current design we should keep it. Current users are already familiar with it. And, it is better than having it at the top where users need to reach for it. 

If we're deciding whether to add a FAB, I would have been against it, as like you said, it is unnecessary with the action mode. But having an already implemented FAB. Why get rid of the work just to regress to a poorer UX?

I vote to keep it.

ByteHamster commented:
> How about adding a 3-dot like button next to the play button.

I think this will clutter the screen too much. Please use the long-press menu for now. Changing that later would be pretty easy, so I would rather go with as few muscle-memory breaking changes for now.

> Regarding the FAB. I think with our current design we should keep it. Current users are already familiar with it. And, it is better than having it at the top where users need to reach for it.

Okay :+1: I think it will be slightly harder to implement than the ActionMode but I agree :)

peakvalleytech commented:
After thinking it over, I agree that having it in the context menu wouldn't be such a bad idea. So let's do that. 

Thanks, it actually isn't too hard. I've already implemented it, and works it works nicely.

peakvalleytech commented:
@ByteHamster i am ready for you to review my latest commit. The multi-select is not working on FeedItemListFragment. If you approve of it, I will proceed with migrating the rest of the views that use episode multi-select. 
To reduce code, I suggest that all fragments that use episode multi-select should have common base classes containing the multi-select logic. Is that ok with you?

ByteHamster commented:
> To reduce code, I suggest that all fragments that use episode multi-select should have common base classes containing the multi-select logic. Is that ok with you?

This makes it harder to modularize and test the code. I think it would be better to do it the other way around: Extract the common parts (like ActionBinding) to their own classes. Then these things can be included in all classes that need them. They can then also be tested independently of the UI.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/FeedItemlistFragment.java:
```diff
@@ -176,13 +242,21 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
             protected void doTint(Context themedContext) {
                 toolbar.getMenu().findItem(R.id.sort_items)
-                        .setIcon(AppCompatDrawableManager.get().getDrawable(themedContext, R.drawable.ic_sort));
+                        .setIcon(R.drawable.ic_sort);
```
Are you sure that this works? The icon color should change when scrolling in the light theme. (White when on the blurred background, black when on the toolbar).

peakvalleytech replied:
I had to make this change because I one of your recents commits removed the black and white icons that are used here. I was going to create a separate issue later.

ByteHamster replied:
On my device, the AppCompatDrawableManager performs the color changes automatically. The images reference an attr resource internally.

peakvalleytech replied:
That's not the problem. The problem is that the ide cannot find the icons. One of your recent commits removed the attr referfence.

ByteHamster replied:
The original/removed line here does not mention an attr reference

peakvalleytech replied:
@ByteHamster Ok, fixed.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/FeedItemlistFragment.java:
```diff
@@ -594,5 +797,29 @@ public FeedItemListAdapter(MainActivity mainActivity) {
+    private void close(@PluralsRes int msgId, int numItems) {
+        ((MainActivity) getActivity()).showSnackbarAbovePlayer(
+                getResources().getQuantityString(msgId, numItems, numItems), Snackbar.LENGTH_LONG);
+        getActivity().getSupportFragmentManager().popBackStack();
```
What does this do on the episodes fragment? Is there anything that needs to be closed?

peakvalleytech replied:
@ByteHamster It was pulled from EpisodesApplyActionFragment which is used for multi select but will no longer be needed after this PR. Due to the context mode a fragment no longer needs closing. Renaming to 'showMessage'

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/view/EpisodeItemListRecyclerView.java:
```diff
@@ -81,4 +89,15 @@ public boolean isScrolledToBottom() {
+
+    }
+
+    public void setChoiceMode() {
```
What does this do?

ByteHamster commented on app/src/main/res/menu/episodes.xml:
```diff
@@ -20,6 +20,7 @@
     <item
         android:id="@+id/filter_items"
         android:icon="@drawable/ic_filter"
+        android:iconTint="?action_icon_color"
```
Why this change?

ByteHamster commented on app/src/main/res/menu/episodes_apply_action_options.xml:
```diff
@@ -9,17 +9,17 @@
         <menu>
             <item android:id="@+id/sort_title_a_z"
-                  android:title="@string/sort_title_a_z"/>
+                android:title="@string/sort_title_a_z"/>
```
Please don't reformat the file, unless you really need to change it. We don't have clear guidelines for xml formatting currently, so it should just be kept as it is.

ByteHamster replied:
Still unrelated changes here :)

ByteHamster commented on core/src/main/res/values/attrs.xml:
```diff
@@ -12,4 +12,5 @@
     <attr name="filter_dialog_clear" format="color"/>
     <attr name="filter_dialog_button_background" format="reference"/>
     <attr name="seek_background" format="color" />
+    <attr name="icon_color_inverted" format="color" />
```
Unused?

peakvalleytech commented:
I'll take a look at the comments soon. Thanks for the reply.

keunes commented:
I'm downloading a bunch of 'debug' artifects to try out - let me know when I can/should do that for this PR :)

peakvalleytech commented:
> I'm downloading a bunch of 'debug' artifects to try out - let me know when I can/should do that for this PR. Eager to try it out :)

I should have something ready soon.

peakvalleytech commented on app/src/main/java/de/danoeh/antennapod/fragment/FeedItemlistFragment.java:
```diff
@@ -106,6 +115,32 @@
+     *
+     * Includes: the menu / action item and the actual logic
+     */
+    public static class ActionBinding {
```
I've studied this part in more detail and I've concluded that this is really just unneeded complexity.  Whoever wrote this was trying to implement good architecture. And while this approach would work in another project, it doesn't work for ours. It deviates from the rest of the code that only requires a switch statement. I say we make it consistent and replace this with a switch statement just like the rest of the code.

ByteHamster replied:
I think it makes sense to have it as a different class. Otherwise, the same `switch` has to be copied to other screens like queue, downloads, etc.

This abstraction also makes it easier to test the code.

peakvalleytech replied:
Well, it don't undestand why this requires testing. It is just a common switch statment for handinling menu click events. 
The code in the click listenrs is just a database call and a call to show a message and exit the multi select mode. Not considering strictly UI related code, it is basically only a database call which we likely already have tests for. 

About other the other screens. I suggest we use a base fragment. That way they can share common logic like the switch.

ByteHamster replied:
> About other the other screens. I suggest we use a base fragment. That way they can share common logic like the switch.

That's exactly why I think that a class for the ActionBinding is useful. When pushing everything to a parent fragment, future developers will tend to add more and more code to that parent fragment, making it some kind of "god class".

> Well, it don't undestand why this requires testing.

For screens with dynamic loading, the adapter does not know about all items. When someone used "select all", we have to apply the change even to items that are not loaded yet. This is currently not a problem because the screens that support dynamic loading simply don't use multi-select but for better performance, this will be needed in the future. It's not the scope of this PR but it will make future changes easier.

peakvalleytech replied:
> That's exactly why I think that a class for the ActionBinding is useful. When pushing everything to a parent fragment, future developers will tend to add more and more code to that parent fragment, making it some kind of "god class".

Ok, then i'll create the Actioninding class.

peakvalleytech replied:
> For screens with dynamic loading, the adapter does not know about all items. When someone used "select all", we have to apply the change even to items that are not loaded yet. This is currently not a problem because the screens that support dynamic loading simply don't use multi-select but for better performance, this will be needed in the future. It's not the scope of this PR but it will make future changes easier.

I'm not sure what you mean. How does this relate to testing.

ByteHamster replied:
Sorry. I meant that the features will likely need to be enhanced and will get more complex. Then testing is useful.

peakvalleytech replied:
@ByteHamster Oh right. So how should I proceed with this PR?

peakvalleytech commented on app/src/main/java/de/danoeh/antennapod/adapter/EpisodeItemListAdapter.java:
```diff
@@ -30,16 +40,25 @@
     private FeedItem selectedItem;
-
+    protected Set<FeedItem> checkedItems = new HashSet<>();
+    final ActionMode[] actionMode = {null};
```
It is the only way I could reference it in the listener. Without it, the compile complains.

ByteHamster replied:
That's probably because you set it to `final`

peakvalleytech commented:
Hey @ByteHamster,
I just pushed a new commit. Still waiting for your response. Hope all is well.

ByteHamster commented:
> Still waiting for your response. Hope all is well.

Sorry, I can't check AntennaPod every day (it's done in my free time). Could you please clean up some of the code here? It contains a lot of commented code and changes to unrelated files. That would make it a lot easier to review.

I just played around with it. It does feel pretty nice. There seems to be a problem though, where wrong episodes can get changed. Also, pressing the "Select all" button twice brings everything into a broken state that always affects all episodes, regardless of what you select.

Other issues I found:

- The ActionBar icons are in fact wrong in the light theme (there is already a comment about this above)
- The items no longer have a Ripple effect when they are normally pressed. I think selection would be easier to handle using a state-list-drawable
- The selected color is a bit too strong, in my opinion. Maybe @keunes has an idea for what color to use instead of the accent blue. 
- When pressing the back button, the ActionMode closes but the FAB stays

peakvalleytech commented:
> Sorry, I can't check AntennaPod every day (it's done in my free time). Could you please clean up some of the code here? It contains a lot of commented code and changes to unrelated files. That would make it a lot easier to review.

Ok that is fine. I'll go and send a new commit soon.

keunes commented:
I'll have a look at the colour later (probably over the weekend)

peakvalleytech commented:
> I'll have a look at the colour later (probably over the weekend)

@keunes have you had the chance to look at this PR yet?

ByteHamster commented:
I was thinking maybe we could use the action button to indicate selection state. When an episode is not selected, just keep it empty. When an episode is selected, display a check mark (blue copy of the drawable). Then we don't have problems with the contrast when setting a backgroubd. Also, the action button still showing "download" could be confusing otherwise.

peakvalleytech commented:
> I was thinking maybe we could use the action button to indicate selection state. When an episode is not selected, just keep it empty. When an episode is selected, display a check mark (blue copy of the drawable). Then we don't have problems with the contrast when setting a backgroubd. Also, the action button still showing "download" could be confusing otherwise.

Yeah that will work nicely. So should I proceed with porting it to the other screens (Queue, Downloads, ... etc.)? Let me know if you need any more changes, especially related to the architecture.

peakvalleytech commented:
@ByteHamster Actually, after looking at the code, I don't think the actionbutton can be changed easilty. I suggest hiding it while in multi select mode. 

I think we can make it work if we change the color. Maybe to a light gray for light theme and dark gray for dark theme.

That is how it is usally done by other apps. It will be better UX.

If you want to still do the checkmark approach. I can add a checkmark view to the EpisodeItemViewHolder. Would that be ok with you?

keunes commented:
> Yes. The idea of using the same layout for both normal interaction and multi-select is consistency. Users don't have to learn to deal with an additional layout.

I think hiding the action buttons is really a different layout. If anything, it's just the same layout with more space for already displayed information (podcast/episode titles). Plus, it could well be reused for the Inbox list item.

As for the colour, @peakvalleytech, I'd [propose](https://coolors.co/2c66c2-d4e3f7-ffffff-757575-4b4b4b-212121)
* Use #d4e3f7 for the background (the blue used for the NEW label, but with luminance cranked up to 90)
* Use #4b4b4b for the metadata text colour (in the middle between the episode title and regular metadata colour) in order to maintain equal contrast ratio

peakvalleytech commented:
> Hmm, changing the text color as well sounds like pretty much "manual" work that the Android system cannot handle for us. Maybe we can live with a slightly lower contrast here.

I already implemented your idea of using checkboxes sorta. Like I said, I don't think it is good design to mess with the action button. That is a button. And what we need is a checkbox. It would be a mixing of concerns which is bad. Makes the code more sloppy and less understandle. 

I instead added a radiobutton to the layout. I already tested to make sure it doesn't break any of the other screens. It doesn't. It is ONLY visible when action mode is active. So I think this is the best approach. It's a good separation of concerns. And it looks great to.

keunes commented:
@peakvalleytech I must say it works & looks nice. A few comments:
* I'm wondering if we should remove the action button. When using multi-select now, it actually ~~reduces the info on screen per episode~~ reduces the space, making list items longer. Removing the action button keeps the space per episode (more or less) consistrn between states.
* While radio buttons look probably nicer than check boxes, they signal that only one option from the list can be selected. For consistency with other apps & platforms I think we should change them to checkboxes.
* I feel that when the last selected option is deselected, we should leave the selection state.
* It's not possible to unselect an episode by tapping on the radio button itself (tapping on the rest of the item does deselect the episode)
* Related: When deselecting the last selected episode, then selecting the episode by tapping on the radio button, doesn't seem to select the item (when you then hit the FAB, a notification is shown that no episodes are selected)

Otherwise: great work! big big improvement compared to current selection state!

peakvalleytech commented:
> @peakvalleytech I must say it works & looks nice. A few comments:
Thanks, not easy. A lot of work.
> * I'm wondering if we should remove the action button. When using multi-select now, it actually ~reduces the info on screen per episode~ reduces the space, making list items longer. Removing the action button keeps the space per episode (more or less) consistrn between states.
Not sure what you mean. Currently it is displayed when not in select mode, and hidden when in select mode.
> * While radio buttons look probably nicer than check boxes, they signal that only one option from the list can be selected. For consistency with other apps & platforms I think we should change them to checkboxes.
Yes, you are correct. I'll them to checkboxes.
> * I feel that when the last selected option is deselected, we should leave the selection state.
That might be a good idea. But if they want to exit, they can just use back button. I think we should have poll. 
> * It's not possible to unselect an episode by tapping on the radio button itself (tapping on the rest of the item does deselect the episode)
That's just extra complexity. It's purpose is to serve as an indicator of item being selected or not. It's not necessary for it to be interactive as the whole item can be selected which is probably the intuitive way (the usual pattern) user uses it.
> * Related: When deselecting the last selected episode, then selecting the episode by tapping on the radio button, doesn't seem to select the item (when you then hit the FAB, a notification is shown that no episodes are selected)
Not sure what you mean. Radio buttons are supposed to be interactive as mentioned above.
> Otherwise: great work! big big improvement compared to current selection state!
Thank you. Hope @ByteHamster feels the same. Would like to proceed with migrating the other screens and get this PR merged. Then we can start work on other issues, like the home screen.

keunes commented:
> That might be a good idea. But if they want to exit, they can just use back button. I think we should have poll.

We could organise a poll with our userbase, but a proper one would slow down the process of finishing this PR - I'm not sure it's worth it.
I was thinking to keep what you made, but _add_ my proposal (last deselect exits multi-select state). If my idea is implemented, the other option you mention (use back button to exit) should still be there. In short: it would be good to have both.

> That's just extra complexity. It's purpose is to serve as an indicator of item being selected or not. It's not necessary for it to be interactive as the whole item can be selected which is probably the intuitive way (the usual pattern) user uses it.

I agree it serves as an indicator. I also agree that tapping on the rest is probably the most intuitive for most people.
But if I (accidentally) tap on it, it's visual state (filled or not filled) is switched, without the selection state actually being switched. At that point the indication is no longer correct/true. Therefore, tapping on the indicator (rather than tapping the rest of the episode), renders the indicator useless.
In the not likely event a user taps the indicator, it should still correctly represent the state. Therefore the selection state should switch when the indicator is tapped. (Does that make sense? Not sure it's clear - it may be best to try yourself what I tried to describe.)

> Not sure what you mean. Radio buttons are supposed to be interactive as mentioned above.

[This description of the difference](https://www.nngroup.com/articles/checkboxes-vs-radio-buttons/) between radio buttons (round) and checkboxes (squared) is written by a well-respected organisation of UX experts. Point 1 & 2 are relevant here, the rest less so. [This article](https://www.justinmind.com/blog/checkboxes-or-radio-buttons-let-the-ui-design-battle-commence/) also gives some nice examples in practice.
In short: round buttons (radio buttons) are only for when you can only select one option. Squares (checkboxes) are for when you can select multiple options. That is a widely accepted standard, and we should follow.
Hope that helps?

ByteHamster commented:
> Thank you. Hope @ByteHamster feels the same.

I definitely think that the UI is an improvement to the old one :) Still needs a few tweaks here and there but I like it.

> Would like to proceed with migrating the other screens and get this PR merged.

I will have a closer look at the code soon. First few things I noticed while skimming the code:

- There are some files that just have whitespace changes and no actual code changes. Please revert those to keep the untouched files' git history clean
- Looks like the actions like `ACTION_ADD_TO_QUEUE` are not used
- Both the classes `EpisodeItemListAdapter` and `SelectableAdapter` keep a list of selected items. This belongs into `SelectableAdapter` only

> I feel that when the last selected option is deselected, we should leave the selection state.

+1.

> I also agree that tapping on the rest is probably the most intuitive for most people.

I don't really agree that tapping next to the checkbox is more intuitive than tapping the checkbox itself. Anyway, as keunes wrote, please do make the checkbox trigger the selection state, too.

peakvalleytech commented:
> > Thank you. Hope @ByteHamster feels the same.
> 
> I definitely think that the UI is an improvement to the old one :) Still needs a few tweaks here and there but I like it.
Yes, i agree. 
> * There are some files that just have whitespace changes and no actual code changes. Please revert those to keep the untouched files' git history clean
Ok.
> * Looks like the actions like `ACTION_ADD_TO_QUEUE` are not used
Yes, i'll clean up this too.
> * Both the classes `EpisodeItemListAdapter` and `SelectableAdapter` keep a list of selected items. This belongs into `SelectableAdapter` only
The selectableAdapter is supposed to be able to work with any class not just FeedItems. It only keeps a list of  id's.
The EpisodeItemListAdapter, which is a subclass of SelectableAdapter keeps track of FeedItem objects. That way
SelectableAdapter can also be reused for subscriptions.
> > I feel that when the last selected option is deselected, we should leave the selection state.
> +1.
Ok, then i'll add that too.
> > I also agree that tapping on the rest is probably the most intuitive for most people.
> I don't really agree that tapping next to the checkbox is more intuitive than tapping the checkbox itself. Anyway, as keunes wrote, please do make the checkbox trigger the selection state, too.
Well, if it normally is if there were no checkbox. But the checkbox does change things. And thinking it over, it actually isn't optional to not make the checkbox also selectable. That would be strange behavior to have it in a checked state yet not included in the selected items.

keunes commented:
> would be strange behavior to have it in a checked state yet not included in the selected items.

I agree. But this is exactly what can happen now, so this needs to be corrected.

Can you download the debug build and try the following? It might help to clarify what I mean is wrong:
* Select an episode
* Tap the radio button of another episode (it turns blue and it _seems_ like it has the selected state)
* Use the FAB button to apply an action (e.g., mark as read)

You will notice that the action you chose did not get applied to the episode on which you tapped the indicator, even though the app gave the impression it was selected.
(The snackbar also stated that the action was applied to one episode, though two episodes gave the impression to be selected.)

peakvalleytech commented:
@keunes
Yes, that is basically what I said: incorrect behavior. I will correct this in the next update.

keunes commented:
>  Yes, that is basically what I said: incorrect behavior. I will correct this in the next update.

@peakvalleytech Ah, ok, sorry. I understood the first half of your comment as an explanation why current behaviour is correct. But all good, then.
Side story: My boyfriend just out of the blue mentioned the multi-select as something that should be improved in AntennaPod. Showing him the debug build he said that was exactly what he wanted :)

peakvalleytech commented:
> > Yes, that is basically what I said: incorrect behavior. I will correct this in the next update.
> 
> @peakvalleytech Ah, ok, sorry. I understood the first half of your comment as an explanation why current behaviour is correct. But all good, then.
> Side story: My boyfriend just out of the blue mentioned the multi-select as something that should be improved in AntennaPod. Showing him the debug build he said that was exactly what he wanted :)

No worries. Boyfriend, eh?. I thought you were male. Good to see contributions from women.

keunes commented:
> Boyfriend, eh?. I thought you were male. Good to see contributions from women.

This is completely irrelevant for the project, but indeed I am male.

peakvalleytech commented:
> > Boyfriend, eh?. I thought you were male. Good to see contributions from women.
> 
> This is completely irrelevant for the project, but indeed I am male.

Well, i'm alright with that. That is good too.

peakvalleytech commented:
@ByteHamster still waiting for a review :)

ByteHamster commented:
Sorry, I don't have a lot of time recently. Just tried your PR again. Feels pretty nice!

I did not look at the code yet but some things I noticed:

- There are still unrelated code changes in the diff, as well as unused code. Please remove that, so that the PR becomes easier to review.
- I wonder if it would be better to display the checkbox on the right instead of the episode action. For some podcasts, starting selection mode currently moves around the whole screen because the checkbox takes away space that then leads to titles being displayed in two rows instead of one. The action button is not useful in multi-select mode anyway.
- The previous selection screen had an option to long-press and choose "select all below" or "select all above". I think it would be great if that option was still available (only when in selection mode).
- Not 100% sure but I think I would rename the context menu entry from "Multi select" to just "Select" and move it to the top.

peakvalleytech commented:
> Sorry, I don't have a lot of time recently. Just tried your PR again. Feels pretty nice!

Thanks. No worries. I haven't had much time either. So we're in the same boat! :)
 
> * There are still unrelated code changes in the diff, as well as unused code. Please remove that, so that the PR becomes easier to review.
I thought I got them all. I'll do another check.
> * I wonder if it would be better to display the checkbox on the right instead of the episode action. For some podcasts, starting selection mode currently moves around the whole screen because the checkbox takes away space that then leads to titles being displayed in two rows instead of one. The action button is not useful in multi-select mode anyway.
Ok, I'll make a revision soon. I'm still new to the code base and so I'm not aware of all the implications of changes.

> * The previous selection screen had an option to long-press and choose "select all below" or "select all above". I think it would be great if that option was still available (only when in selection mode).

Ok, got it. I didn't implement because I wasn't sure if was still desired. I'll go ahead and add the options.
> * Not 100% sure but I think I would rename the context menu entry from "Multi select" to just "Select" and move it to the top.

Hmm, "Select" seems ambiguous, as they are already selecting in when long pressing. Other apps use "Batch edit".  I think we should have a vote. 
I am, however, in favor of moving it to the top.

ByteHamster commented:
>> I wonder if it would be better to display the checkbox on the right instead of the episode action.
>
> Ok, I'll make a revision soon

I just painted a few mock-ups of the check mark position. Which one do you like most (also @keunes)? :)

<img width="200" src="https://user-images.githubusercontent.com/5811634/119219268-f1626780-bae4-11eb-9ee8-593c8056da06.png" /> <img width="200" src="https://user-images.githubusercontent.com/5811634/119219271-f2939480-bae4-11eb-9558-b3143c697c05.png" /> <img width="200" src="https://user-images.githubusercontent.com/5811634/119219274-f32c2b00-bae4-11eb-9bfc-7d4a419807ad.png" />

peakvalleytech commented:
I just relized that having the checkbox on the right would cause the last checkbox to be covered by the FAB and therefore be unselecteable. Although this isn't very serious as the item can still be selected by clicking on the item itself. 

And I know you pointed out that having it on the left would cause some titles to need more lines but the checkboxes are only shown in select mode and so is the extra lines will be removed once out of select mode. 

To me, I think we should use the first screen with the checkbox on the left. That is already implemented. Means less work and more time on more important parts. I'm not in favor of using the action button as a selection indicator. I say we make it invisible while in select mode. 

@keunes?

ByteHamster commented:
> the checkboxes are only shown in select mode and so is the extra lines will be removed once out of select mode

That's basically my point. Things look pretty different between select mode and normal mode. At some point in the future, we might want to make select mode the default long-press action - like in other apps. Then, whenever you want to mark something as played, the whole screen shifts around. If, instead, the already existing action buttons are used, entering multi-select mode does not cause any movement on the screen.

keunes commented:
I agree with you both, I think :)
* Keep checkbox on the left (good point about the FAB @peakvalleytech; even if you can select, you can't see it's status)
* Remove the action buttons (as @ByteHamster said they're irrelevant)

> If, instead, the already existing action buttons are used, entering multi-select mode does not cause any movement on the screen.

I think the movement is so small that it doesn't matter. The checkboxes appear very smoothly, it feels nice. If the action buttons disappear in the same way at the same time, I feel it not be distracting. (Also because the effective width stays the same, so texts/items don't get longer/shorter.)

peakvalleytech commented:
> That's basically my point. Things look pretty different between select mode and normal mode. At some point in the future, we might want to make select mode the default long-press action - like in other apps. Then, whenever you want to mark something as played, the whole screen shifts around. If, instead, the already existing action buttons are used, entering multi-select mode does not cause any movement on the screen.

That's a good point. Thanks for clarifying. I'll give your suggestion a shot.

ByteHamster commented:
Visually, I do like the current version (middle screenshot) most but I think it might be better for the future to try to keep the list items as consistent as possible between modes. Not completely sure what we should use. I asked two people - one (a user) preferred the action buttons (right) and the other person (no user) preferred the checkboxes (middle).

peakvalleytech commented:
> I agree with you both, I think :)
> 
> * Keep checkbox on the left (good point about the FAB @peakvalleytech; even if you can select, you wouldn't be able to see the last item's status)
> * Remove the action buttons (as @ByteHamster said they're irrelevant)
> 

> I think the movement is so small that it doesn't matter. The checkboxes appear very smoothly, it feels nice. If the action buttons disappear in the same way at the same time (so that would at this point require extra work I think?), I feel it will not be distracting. (Also because the effective width stays the same, so texts/items don't get longer/shorter.)

So the main problem seems to be the shifting of the layout or having the last checkbox hidden. 

I willing to do whatever @ByteHamster feels is best.

peakvalleytech commented:
> Visually, I do like the current version (middle screenshot) most but I think it might be better for the future to try to keep the list items as consistent as possible between modes. Not completely sure what we should use. I asked two people - one (a user) preferred the action buttons (right) and the other person (no user) preferred the checkboxes (middle).

How about we stick to the middle (for now). And once we get more data about user's preferences we can decide to change it or not.

peakvalleytech commented:
@ByteHamster, hey I know you said you you were busy, but I would like you to review my latest update to this PR. That way I can proceed with finalizing the last iteration. 

Also, once we get this done, we can start working on adding multi select to subscriptions. ;)

ByteHamster commented:
This PR is making great progress :) Using it feels pretty nice :)

> How about we stick to the middle (for now).

The height of list items still changes, even if the action buttons are now hidden. See video: https://user-images.githubusercontent.com/5811634/120904193-34771b80-c64b-11eb-99b0-fc09b2ec8634.mp4

Some things I noticed when playing around with it:

- Just that it doesn't get forgotten: The multi-select option in the 3-dot overflow menu should be removed, now that we have the integrated screen.
- "select all below" selects all below, **including** the episode that you pressed. "select all above" selects **excluding** the pressed episode. I think it should be consistent.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/adapter/EpisodeItemListAdapter.java:
```diff
@@ -5,43 +5,48 @@
     private List<FeedItem> episodes = new ArrayList<>();
+    // Used to enable O(1) search. Can't use FeedItem because it does not override equals()
+    private Set<Long> selectedIds = new HashSet<>();
+    private List<FeedItem> selectedItems = new ArrayList<>();
```
Oh, that's the reason why you have two different data structures here. Does the order of the `selectedItems` list matter? If not, I think it would be more clear if both would be a Set.

peakvalleytech replied:
My original intent was to use a set of FeedItem objects. But that is not possible since it doesn't override equals(). Hence, the need to use selectedIds.

ByteHamster replied:
You can use the adapter's `getItemId()` from `SelectableAdapter` to get an ID that identifies an item (for showing selected state). I would then build the list of `FeedItem`s only when actually handling the touch action. That way, there is one single location that keeps track of the selection (so nothing can go out of sync).

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/adapter/EpisodeItemListAdapter.java:
```diff
@@ -5,43 +5,48 @@
+    // Used to enable O(1) search. Can't use FeedItem because it does not override equals()
+    private Set<Long> selectedIds = new HashSet<>();
+    private List<FeedItem> selectedItems = new ArrayList<>();
     private FeedItem selectedItem;
+    int selectedPosition = 0; // used to init actionMode
```
Maybe another name here? "Select" now is used for both, the item that opened the context menu and the list of checked items.

peakvalleytech replied:
Changed to initialItemPosition. Also added another comment for further clarification.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/adapter/EpisodeItemListAdapter.java:
```diff
@@ -136,11 +202,37 @@ protected Activity getActivity() {
+        void onEndSelectMode();
+    }
+
+    private void selectHandler(int pos) {
```
How about calling it `toggleSelection` and moving it to `SelectableAdapter`?

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/adapter/SelectableAdapter.java:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    private ActionMode actionMode;
+    private SparseBooleanArray selectedItemPositions = new SparseBooleanArray();
```
Doesn't this serve the same purpose as the `selectedIds` list in the child class? I think it would be better if keeping track of which (and how many) items are selected would be tracked in only one single location. Maybe using the RecyclerView method `getItemId()`?

I am a bit worried about using positions. What happens if the podcast gets refreshed and has a new episode while in multi-select mode? Then the positions shift around and I think it could break things. Storing IDs should me more stable than positions.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/adapter/SelectableAdapter.java:
```diff
@@ -0,0 +1,237 @@
+    public SelectableAdapter(Activity activity) {
+        this.activity = activity;
+    }
+    @NonNull
+    @Override
+    public T onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+        return null;
+    }
+
+    @Override
+    public void onBindViewHolder(@NonNull T holder, int position) {
+    }
+
+    @Override
+    public int getItemCount() {
+        return 0;
+    }
```
If you make the class abstract, you don't need to provide stub implementations

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/adapter/SelectableAdapter.java:
```diff
@@ -0,0 +1,237 @@
+
+    public void startActionMode(int pos) {
+        if (inActionMode()) {
+            finish();
```
Could be misleading because finish in activities closes the page. How about stopActionMode?

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/adapter/SelectableAdapter.java:
```diff
@@ -0,0 +1,237 @@
+        }
+    }
+
+    public void onStartActionMode() {
+
+    }
+
+    protected void onEndActionMode() {
+
+    }
```
Abstract?

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/adapter/SelectableAdapter.java:
```diff
@@ -0,0 +1,237 @@
+     * @param pos the position that was changed
+     * @param selected the new state of the selection, true if selected, otherwise false
+     */
+    public void onSelectChanged(int pos, boolean selected) {
```
Abstract?

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/FeedItemlistFragment.java:
```diff
@@ -106,13 +114,22 @@
     private ToolbarIconTintManager iconTintManager;
+
+    private SpeedDialView mSpeedDialView;
+    private  LongList checkedIds = new LongList();
```
Oh, another copy of the selected items? Please pass them to `EpisodeMultSelectActionHandler` as an argument instead.

peakvalleytech replied:
Ok, yeah, you're correct. That is a good idea. I've extracted the checkedIds to EpisodeMultSelectActionHandler.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/FeedItemlistFragment.java:
```diff
@@ -323,6 +375,27 @@ public boolean onContextItemSelected(@NonNull MenuItem item) {
+            mSpeedDialView.setVisibility(View.VISIBLE);
+            refreshToolbarState();
+            return true;
+        }
+
+        if (item.getItemId() == R.id.select_all_above) {
```
Shorter:

```suggestion
        } else if (item.getItemId() == R.id.select_all_above) {
```

ByteHamster commented on app/src/main/res/layout/feed_item_list_fragment.xml:
```diff
@@ -76,5 +76,33 @@
             android:layout_gravity="bottom"
             android:visibility="gone"/>
+    <com.leinardi.android.speeddial.SpeedDialOverlayLayout
+        android:id="@+id/fabSDOverlay"
```
Please use the same indention as the rest of the file

ByteHamster commented on app/src/main/res/layout/feeditemlist_item.xml:
```diff
@@ -43,6 +43,12 @@
                     tools:background="@android:color/holo_green_dark"/>
 
+            <CheckBox
+                android:id="@+id/selectCheckBox"
```
Please use the same indention as the rest of the file

ByteHamster reviewed (commented):
I am happy with the overall code structure now but I have yet another round of small comments for you ;)

Now that the structure looks good, it would be great if you could also start making the Checkstyle CI test happy.

peakvalleytech commented:
> I am happy with the overall code structure now but I have yet another round of small comments for you ;)
> 
> Now that the structure looks good, it would be great if you could also start making the Checkstyle CI test happy.

Ok, i haven't seen the code in a while. I'll have to refresh myself. Once I do i'll go through your comments and start finalizing the design.

ByteHamster commented:
> Ok, i haven't seen the code in a while. I'll have to refresh myself

Sorry, I am pretty slow with my replies recently. So no worries if addressing the feedback takes a bit longer :) Once we have a great implementation for the podcast screen, it should be easy (and lead to much fewer discussions) to add it everywhere else :)

peakvalleytech commented:
> > Ok, i haven't seen the code in a while. I'll have to refresh myself
> 
> Sorry, I am pretty slow with my replies recently. So no worries if addressing the feedback takes a bit longer :) Once we have a great implementation for the podcast screen, it should be easy (and lead to much fewer discussions) to add it everywhere else :)

I understand. I admire developers like you who run OSS. It is a lot of effort for not a lot if any compensation. The fact that you still put in as much effort on a regular basis is amazing. 

I didn't have much time to work on it yesterday. I was mostly getting reacquainted with the code. Today, I should start making the changes.

Regarding some of the issues you mentioned. A lot of what I wrote is with the intent that it would be reused for other fragments that will have multi-select (not just for episodes, but soon subscriptions as well).

ByteHamster commented on app/src/main/res/menu/feeditemlist_context.xml:
```diff
@@ -1,6 +1,12 @@
 
 <menu xmlns:android="http://schemas.android.com/apk/res/android">
+    <item
+        android:id="@+id/episode_actions"
```
Please call it `multi_select` instead. `episode_actions` does not tell readers that it is about selecting multiple items, not about actions for a single item.

ByteHamster commented on app/src/main/res/layout/feeditemlist_item.xml:
```diff
@@ -43,6 +43,12 @@
+                    android:id="@+id/selectCheckBox"
+                    android:layout_width="0dp"
+                    android:layout_height="match_parent"
+                    android:layout_weight="1" />
```
What is the weight for? Weights are used when displaying multiple items in a single layout but we never actually want to display both drag handle and checkbox.

ByteHamster commented on app/src/main/res/layout/feed_item_list_fragment.xml:
```diff
@@ -76,5 +76,33 @@
+                android:visibility="gone"
+                app:sdMainFabClosedSrc="@drawable/ic_fab_edit"
+                app:sdOverlayLayout="@id/fabSDOverlay" />
+    </ScrollView>
```
This will probably be used in quite a few other fragments as well. Please create a new file for this and use a `<merge>` tag to include it here.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/view/EpisodeItemListRecyclerView.java:
```diff
@@ -5,10 +5,12 @@
 import androidx.recyclerview.widget.DividerItemDecoration;
 import androidx.recyclerview.widget.LinearLayoutManager;
 import androidx.recyclerview.widget.RecyclerView;
+
```
Please revert changes to unrelated file

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/FeedItemlistFragment.java:
```diff
@@ -585,7 +656,7 @@ private Feed loadData() {
     }
 
-    private static class FeedItemListAdapter extends EpisodeItemListAdapter {
+    private class FeedItemListAdapter extends EpisodeItemListAdapter {
```
Why?

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/FeedItemlistFragment.java:
```diff
@@ -558,16 +629,16 @@ private void loadItems() {
+                            refreshHeaderView();
+                            displayList();
+                            Log.e(TAG, Log.getStackTraceString(error));
+                        });
```
If you completely revert the unrelated whitespace formatting, Checkstyle should not complain. It only complains about lines that you changed.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/FeedItemlistFragment.java:
```diff
@@ -323,6 +363,25 @@ public boolean onContextItemSelected(@NonNull MenuItem item) {
+            mSpeedDialView.setVisibility(View.VISIBLE);
+            refreshToolbarState();
+            return true;
+        }
+
+        if (item.getItemId() == R.id.select_all_above) {
```
```suggestion
        } else if (item.getItemId() == R.id.select_all_above) {
```

ByteHamster commented:
I just removed 2 of the 3 lists that store which episodes are selected and also did a bit of cleanup. I think I am happy with the code structure now. Please have another look again to see if I overlooked something.

peakvalleytech reviewed (commented):
> I just removed 2 of the 3 lists that store which episodes are selected and also did a bit of cleanup. I think I am happy with the code structure now. Please have another look again to see if I overlooked something.

@ByteHamster I looked at the code a bit, but haven't tested it. I didn't code it like that on a whim. This was the result of a lot of manual testing. Let's hope you didn't break anything. :) After I test it, i'll give you an update.

peakvalleytech commented:
@ByteHamster Ok, i've reviewed your changes. And I am also happy with the code structure. 

Now that we have implemented it for FeedItemListFragment. We just need to implement it for the rest of the screens that need episode multi select. 

I believe those are (at least those that currently have multi-select)
CompletedDownloadsFragment
QueueFragment

Should we merge and close this PR or wait till we have it implemented for the rest of the screens?
I think we should merge this and open a new PR for better organization.

ByteHamster commented:
> I think we should merge this and open a new PR for better organization.

Yeah

I just noticed a little problem before merging this. All screens (eg playback history) now show the multi-select context menu item, even if they don't support multi select.

peakvalleytech commented:
> > I think we should merge this and open a new PR for better organization.
> 
> Yeah

Ok, once we get this merged we can start work on implementing it for the other fragments.
These are QeueFragment and CompletedDownloadsFragment, right?

> I just noticed a little problem before merging this. All screens (eg playback history) now show the multi-select context menu item, even if they don't support multi select.
I'll fix this today.

ByteHamster commented:
Your most recent commit breaks long-pressing in select mode (for "select all above")

ByteHamster commented:
Thanks :)

peakvalleytech commented:
> Thanks :)

My pleasure. Now to start on migrating the other screens to use this multi select. Any suggestions on how this should be done?

The two other screens that use multi select are the QeueFragment and thd CompletedDownloadsFragment. I know the UI is about to under go major changes so I want to avoid doing work that will end up being thrown away anyway. The only screen I know for certain will not be changed is the Queue. Should i start with that?
